### PR TITLE
Fixes event ordering for `onLayout`

### LINKED ISSUE #10421: Fire onLayout in depth-first order
### Problem Description

Other RN platforms fire the `onLayout` event in depth-first order. See, for example, this code in the Android Paper UIManager: https://github.com/facebook/react-native/blob/2c5a966054c2a80d63aa9c2c2a234a80a02918cf/ReactAndroid/src/main/java/com/facebook/react/uimanager/UIImplementation.java#L968-L974

One example where this can cause issues is likely in VirtualizedList, which could presume that the list items fire `onLayout` before the scroll container fires `onLayout`.

### Steps To Reproduce

Pretty simple to confirm on other platforms. Create an app like:
```
<View onLayout={() => console.log('Layout Parent')}>
  <View onLayout={() => console.log('Layout Child')} />
</View>
```

Observe that the event order is different on RN Windows than other platforms.

### Expected Results

Event ordering should be consistent across platforms.

### CLI version

npx react-native --version

### Environment

```markdown
npx react-native info
```

### Target Platform Version

_No response_

### Target Device(s)

_No response_

### Visual Studio Version

_No response_

### Build Configuration

_No response_

### Snack, code example, screenshot, or link to a repository

_No response_

### PR DESCRIPTION
**Pull request by rozele**
### Type of Change
_Erase all that don't apply._
- Bug fix (non-breaking change which fixes an issue)

### Why
Other platforms fire `onLayout` depth first. Windows should do the same. Also, once we upgrade to React 18, it is a common case that React commits will create nodes that are not yet attached to any root (and thus do not participate in the Yoga layout for that batch / commit). This means that, in the current approach for updating layout, we would fire `onLayout` events with `NaN` values for width and height.

Resolves #10421

### What
This changes switches from iterating over all Yoga nodes to update layout on each UIManager batch to recursively updating layout from the root shadow node. The effect is that we fire events depth first and we only attempt to update layout on nodes that have participated in Yoga layout (because they are descendants of some root).

This also likely fixes a bug in our app we were seeing where attempt to transfer an explicit NaN Width/Height property when replacing a ViewPanel with a ViewControl or Border (for relevant prop implementations) would occasionally throw an exception.

## Testing
Ran RNTester, everything still seems to work as expected.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/10422)

### CODE REVIEW

**Review by jonthysell — APPROVED** [review]

File: vnext/Microsoft.ReactNative/Modules/NativeUIManager.cpp [hunk 1]
```diff
@@ -939,28 +939,31 @@ void NativeUIManager::DoLayout() {
...
-      float width = YGNodeLayoutGetWidth(yogaNode);
-      float height = YGNodeLayoutGetHeight(yogaNode);
-
-      ShadowNodeBase &shadowNode = static_cast<ShadowNodeBase &>(m_host->GetShadowNodeForTag(tag));
-      auto view = shadowNode.GetView();
-      auto pViewManager = shadowNode.GetViewManager();
-      pViewManager->SetLayoutProps(shadowNode, view, left, top, width, height);
-    }
+void NativeUIManager::SetLayoutPropsRecursive(int64_t tag) {
+  SystraceSection s("NativeUIManager::DoLayout::SetLayoutProps");
```
Comment (jonthysell) [inline vnext/Microsoft.ReactNative/Modules/NativeUIManager.cpp:948]: Should this SystraceSection be here, or up above in the initial calling function?

Comment (rozele) [reply vnext/Microsoft.ReactNative/Modules/NativeUIManager.cpp:948]: Good catch.

File: vnext/Microsoft.ReactNative/Modules/NativeUIManager.cpp [hunk 2]
```diff
@@ -939,31 +939,36 @@ void NativeUIManager::DoLayout() {
...
+    {
+      SystraceSection s("NativeUIManager::DoLayout::SetLayoutProps");
+      SetLayoutPropsRecursive(rootTag);
     }
   }
 }
 
+void NativeUIManager::SetLayoutPropsRecursive(int64_t tag) {
+  ShadowNodeBase &shadowNode = static_cast<ShadowNodeBase &>(m_host->GetShadowNodeForTag(tag));
+  for (const auto child : shadowNode.m_children) {
```
Comment (rozele) [inline vnext/Microsoft.ReactNative/Modules/NativeUIManager.cpp:952]: Check for measure function here to prevent unnecessary recursion.

**Comment by rozele** [comment]
Based on convo with @asklar, moving to draft until we confirm that this does not break rn-xaml.

**Comment by rozele** [comment]
Re-opening, this does not make RN-xaml any worse 😬.

**Pull request merged**